L1Trigger/L1THGCal/src/fe_codecs/HGCalTriggerCellThresholdCodecImpl.cc

```diff
@@ -0,0 +1,205 @@
+            continue;
+        }
+        // else fill result vector with data
+        // (set the corresponding adress bit and fill energy if >0)
+        if(index>=nCellsInModule_) 
+        {
+          throw cms::Exception("BadGeometry")
+            << "Number of trigger cells in module too large for available data payload\n";
+        }
+        uint32_t value = data_itr->second; 
+        if(value>0)
+          {
+            if(idata>=size)
+              {
+                throw cms::Exception("BadData") 
+                  << "encode: Number of non-zero trigger cells larger than codec parameter\n" \
+                  << "      : Number of energy values = "<<size<<"\n";
+              }
+            // Set map bit to 1
+            result[index] =  1;
+            // Saturate and truncate energy values
+            if(value+1>(0x1u<<triggerCellSaturationBits_)) value = (0x1<<triggerCellSaturationBits_)-1;
+            for(size_t i=0; i<dataLength_; i++)
+              {
+                // remove the lowest bits (=triggerCellTruncationBits_)
+                result[nCellsInModule_ + idata*dataLength_ + i] = static_cast<bool>(value & (0x1<<(i+triggerCellTruncationBits_)));
+              }
+            idata++;
+          }
+        index++;
+      }
+    return result;
+}
+
+HGCalTriggerCellThresholdCodecImpl::data_type 
+HGCalTriggerCellThresholdCodecImpl::
+decode(const std::vector<bool>& data, const uint32_t module, const HGCalTriggerGeometryBase& geometry) const 
+{
+    data_type result;
+    result.reset();
+    // TODO: could eventually reserve result memory to the max size of trigger cells
+   
+    HGCalTriggerGeometryBase::geom_ordered_set trigger_cells_in_module = geometry.getOrderedTriggerCellsFromModule(module);
+    size_t iselected = 0;
+    size_t index = 0;
+    for(const auto& triggercell : trigger_cells_in_module)
+    {
+        if(index>=nCellsInModule_)
+        {
+            throw cms::Exception("BadGeometry")
+                << "Number of trigger cells in module too large for available data payload\n";
+        }
+        if(data[index])
+        {
+            uint32_t value = 0;
+            for(size_t i=0;i<dataLength_;i++)
+            {
+                size_t ibit = nCellsInModule_+iselected*dataLength_+i; 
+                if(data[ibit]) value |= (0x1<<i);
+            }
+            iselected++;
+            // Build trigger cell
+            if(value>0)
+            {
+                // Currently no hardware eta, phi and quality values
+                result.payload.emplace_back(reco::LeafCandidate::LorentzVector(),
+                        value, 0, 0, 0, triggercell); 
+                GlobalPoint point = geometry.getTriggerCellPosition(triggercell);
+                // 'value' is hardware, so p4 is meaningless, except for eta and phi
+                math::PtEtaPhiMLorentzVector p4((double)value/cosh(point.eta()), point.eta(), point.phi(), 0.);
+                result.payload.back().setP4(p4);
```

Comment:
An small update has been implemented in the best choice codec, filling in addition the trigger cell position:
```c++
result.payload.back().setPosition(point);
```
This update should be propagated here as well.